conda-store-server/conda_store_server/__init__.py

```diff
@@ -1,6 +1,88 @@
+import datetime
+import typing
 from pathlib import Path
 
 __version__ = "2023.10.1"
 
 
 CONDA_STORE_DIR = Path.home() / ".conda-store"
+
+
+class BuildKey:
```

Comment:
Nice, this looks way better. However I still sniff some code smell here 🤔 I don't get why we are stuck with a multiversion class in the `__init__` module. I would have assumed it was simple enough to have a setup like this:

```python
# optionally, we could define a base class with the common stuff
# omitted for clarity
class BuildKeyV1:
  ...  # v1 specificities
class BuildKeyV2:
  ...  # v2 specificities, overrides as necessary

BUILD_KEY_VERSIONS = {
  1: BuildKeyV1,
  2: BuildKeyV2,
  None: BuildKeyV2,
}
```

This can be all put in a separate module if necessary. Then the consumer code will check in the versions dict which one to use according to the current config. In other words, I think we need to take a look at the problem that causes the circular import and see if there's a better way to define these ideas.

Thanks for the patience! I feel we are getting closer.

Reply:
> I think we need to take a look at the problem that causes the circular import

What we want:
- We want a user-configurable build-key
- The argument/config parsing is done in `conda-store-server/conda_store_server/app.py:CondaStore`
- After reading a config value, we want to update the `orm` module.

Problems:
- The `orm` module is loaded earlier than `app` on startup
- The `app` module also depends on the `orm` module (imports it).

Solution:
- A common module (`__init__.py` in this case) and delayed initialization (`default=_get_build_key_version` in `orm.py`) is a way to solve this
- A common module avoids a cyclic import
- Delayed initialization avoids caching the value on early imports. It's always fetched from a common module after it's been initialized once we actually want to use the `orm` module (as opposed to just importing it).

----

> BUILD_KEY_VERSIONS

I don't understand this new proposal at all.

- What's exactly wrong with the `BuildKey` class?
- What issue does the new proposal solve compared to the existing `BuildKey` class?
- Why do we want to have 2 classes for this? We could, but I just don't see the benefit.

Moreover, I see some problems with this new proposal:
- Nothing prevents me from doing `BUILD_KEY_VERSIONS[non-existent-version]` and getting a generic error. In the current `BuildKey` implementation, error-checking is shared and hidden.
- This will still need another global variable to track the current version (to index into this dict). In the current `BuildKey` implementation, this is hidden inside the class. All `BuildKey`-related things are part of the class.
- Regarding `None: BuildKeyV2`: This is wrong and will hide potential issues if build key version is not initialized (while the user might be requesting BuildKeyV1 in the config). This can happen when a module is imported earlier than the build key version is set and the value is cached. This currently doesn't happen, but someone might run into this issue when refactoring. In my code, I have an assertion to guard against that.

**Summary**:
- I'm going to follow up offline to clarify this proposal
- Its value is not clear to me and I see some problems with it
- Implementing and testing this again would take time away from other things I could be working on.